numba/typeinfer.py

```diff
@@ -882,7 +882,21 @@ def unify(self):
         def check_var(name):
             tv = self.typevars[name]
             if not tv.defined:
-                raise TypingError("Undefined variable '%s'" % (var,))
+                def find_offender(name):
+                    """
+                    finds the offending variable definition to allow its
+                    inclusion in an error message
+                    """
+                    for block in self.func_ir.blocks.values():
```

Comment:
There are some methods under class Block that you can use.  See  https://github.com/numba/numba/blob/b6b965afaf88594c1f754c9db07f3cf73b824478/numba/ir.py#L738-L755

Maybe, this function can be a method in Block, too.

Reply:
I think it's a good idea to have it as a method on `Block`, might help with other error message handling. Will patch.